aff4/aff4_image.cc

```diff
@@ -28,7 +28,7 @@ AFF4Status CompressZlib_(const std::string &input, std::string* output) {
     uLongf c_length = compressBound(input.size()) + 1;
     output->resize(c_length);
 
-    if (compress2(reinterpret_cast<Bytef*>(const_cast<char*>(output->data())),
+    if (compress2(reinterpret_cast<Bytef*>(&(*output)[0]),
```

Comment:
Honestly I think the const_cast (although ugly) is easier to read and reason about than this proposed change.

Reply:
Indeed, this is not my preferred change, as my preferred change would have `output` as a reference instead of a pointer.

Reply:
Incidentally, `std::string::data()` has a non-`const` version in C++17.

Reply:
So why the change in the first place?  If the change is harder to reason about, one should prefer the longer version.  Obviously the same applies for the rest of these changes in this function.

Reply:
My preference would be to not use `std::string` everywhere as a container and instead create a buffer type based on `std::vector<char>`, but that's a large undertaking at this point, and I'm not sure about what @scudette thinks about it.

Reply:
I don't agree that the change is harder to reason about, and casting away `const` strongly suggests something is wrong. (It's not wrong in this case, but it's distracting to see something which should raise red flags and then have to remind yourself each time why in this particular case it's ok.)

Reply:
I should state that more strongly: Casting away `const` means that you can't rely on the compiler to enforce `const`-correctness anymore.

Reply:
> I should state that more strongly: Casting away `const` means that you can't rely on the compiler to enforce `const`-correctness anymore.

In the general case, yes, but in this case that's not relevant.  The crux of this issue is the design choice to use a container that exposes its data as `const char *` instead on one that would expose it as `char *` as intended.

I suppose that it's nice to be able to concatenate the buffers and stuff, and that is done in some places.

Reply:
Also, if you're going that route, you might want to rebase to the latest `master`.  I used the same `const_cast` pattern for the LZ4 compression functions.

Reply:
I got the two from your LZ4 commit, plus one other that I'd missed. Thanks.

Reply:
Just a very general comment that the Google C++ Style Guide has the virtue of being a defined guide, and obviously @scudette it's something you know well, but, at least to outsiders, it's historically reflected a lot of Google's biases against C++ and is not the most welcoming style guide to experienced C++ developers. OTOH, no one wants to write crazy Boost code, either.

Reply:
I agree.  I prefer https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines

Reply:
So I think the main point of contention here is what to do about the const of string.data(). Arguably this is a design fault in the C++ library which is being corrected by c++17:

https://stackoverflow.com/questions/39200665/directly-write-into-char-buffer-of-stdstring/39200666

So I think this kind of pattern &output[0] or &output.front() is kind of an acceptable workaround for this problem and most c++ programmers are used to it being used in this way. I agree that using const_cast is typically a red flag so I think in this case this is actually a good change as it says to the reader - nothing to see here move along :-).

I am not sure how passing by reference vs. passing by pointer makes a difference for this?

Re the style guide - I feel that C++ is a complex and dangerous language without using a very strict and battle tested style guide. There are many subtle bugs introduced by incorrect style. While at Google the Google style guide was obviously used exclusively. These days (luckily) I dont write much C++ so I dont mind either way. 

One of the strengths of the Google guide is that it explicitly forbids certain C++ features. Some people might object - clearly someone introduced those featured into the standard because they thought they were good ideas. At large scale companies though these features have proven to be problematic and tend to introduce many bugs which is why the style guide forbids them (e.g. use of exceptions). Most of the guideline in the Google style guide was born out of a lot of experience writing complex large c++ projects. It may seen biased but this is necessary to tame this large ugly beast that is c++ :-).

Reply:
BTW do you prefer &(*output)[0] or &output->front() ?

The second form is less typing :-).